tests/ui/resolve/issue-111312.rs

```diff
@@ -7,7 +7,6 @@ trait Has {
 trait HasNot {}
 
 fn main() {
-    HasNot::has();
-    //~^ ERROR trait objects must include the `dyn` keyword
-    //~| ERROR no function or associated item named `has` found for trait `HasNot`
+    <dyn HasNot>::has();
```

Comment:
:/ So does this PR undoes the work done by #111588? I'm pretty sure that #111312 was specifically about using `HasNot::has`. This change is not faithful to the original version.

Reply:
I'll see if I can do something in E0782 to at the resolve, but I am generally of the position that if an error leads to another error that has the right suggestion, that is reasonable.

Reply:
Now I check if `HasNot` is object safe and if the type is part of a qualified path to an associated item that didn't resolve (unresolved assoc item can be because it doesn't exist in the trait *or* because it exists but an appropriate self ty couldn't be found). If it is, I don't turn it into `{type error}`, which makes the error about "`has` not found" still be emitted. But if the trait is not object safe, we still turn it into a `{type error}` to avoid any further errors. At the same time, I've added more context to the E0782 about object safety so we don't lose any information when emitting only that one.